contracts/Treasury.sol

```diff
@@ -0,0 +1,173 @@
+        bool isActive;
+    }
+
+    CryptoBlades public game;
+
+    function initialize(CryptoBlades _game) public initializer {
+        __AccessControl_init_unchained();
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _setupRole(GAME_ADMIN, msg.sender);
+
+        game = _game;
+        // each block increases payout by 0.0001x
+        multiplierUnit = 1e4;
+    }
+    
+    modifier restricted() {
+        _restricted();
+        _;
+    }
+
+    function _restricted() internal view {
+        require(hasRole(GAME_ADMIN, msg.sender), "Not game admin");
+    }
+
+    // Views
+
+    function getAmountOfActiveProjects() public view returns (uint256) {
+      uint256 activeCount = 0;
+        for(uint i = 0; i < partneredProjects.length; i++) {
+            if(partneredProjects[i].isActive) {
+              activeCount += 1;
+            }
+        }
+        return activeCount;
+    }
+
+    function getActivePartnerProjectsIds() public view returns (uint256[] memory){
+        uint256 activeCount = getAmountOfActiveProjects();
+        uint256 counter = 0;
+        
+        uint256[] memory activeProjectsIds = new uint256[](activeCount);
+        for(uint i = 0; i < partneredProjects.length; i++) {
+            if(partneredProjects[i].isActive) {
+              activeProjectsIds[counter] = i;
+              counter++;
+            }
+        }
+
+        return activeProjectsIds;
+    }
+
+    function getPartnerProject(uint256 partnerId) public view returns(uint256, string memory, string memory, address, uint256, uint256, uint256, bool) {
+        PartnerProject memory p = partneredProjects[partnerId];
+        uint256 claimed = tokensClaimed[partnerId];
+        return(p.id, p.name, p.tokenSymbol, p.tokenAddress, p.tokenSupply, claimed, p.tokenPrice, p.isActive);
+    }
+
+    function getProjectMultiplier(uint256 partnerId) public view returns(uint256) {
+        return uint(1e18).div(multiplierUnit).mul(block.number.sub(projectAddedBlockNumber[partnerId])).add(1e18);
+    }
+
+    function getProjectClaimedAmount(uint256 partnerId) public view returns(uint256) {
+        return tokensClaimed[partnerId];
+    }
+
+    function getSkillToPartnerRatio(uint256 partnerId) public view returns(int128) {
+        return ABDKMath64x64.divu(skillPrice, partneredProjects[partnerId].tokenPrice);
+    }
+
+    function getRemainingPartnerTokenSupply(uint256 partnerId) public view returns(uint256) {
+        return IERC20(partneredProjects[partnerId].tokenAddress).balanceOf(address(this));
+    }
+
+    function getAmountInPartnerToken(uint256 partnerId, uint256 skillAmount) public view returns(uint256) {
+        uint256 baseAmount = getSkillToPartnerRatio(partnerId).mulu(skillAmount);
+        uint256 bonusAmount = baseAmount.mul(block.number.sub(projectAddedBlockNumber[partnerId])).div(multiplierUnit);
+        return baseAmount.add(bonusAmount);
+    }
+
+    // Mutative
+
+    function addPartnerProject(
```

Comment:
just wanna confirm if we don't expect to be removing projects at some point in the future.
I suspect a lot of projects will have one run with us, then no refill. But i dont know how frequently these will start up.
It really only affects the views in terms of performance, which should still be very managable in the long term, just wanted to think ahead.
May be worth investigating what level of pain comes from trying to remove some projects in the future, since they're stored in an array, and it's not frictionless to remove elements in the middle. But overwriting a slot is still an option.

Reply:
If we wanted to remove them I could probably make new projects to override the first inactive project in array.

On the other hand we could create a second tab on the front-end to show finished (inactive) projects - might be cool to look at, like a sort of history/achievement tab. I think that was my initial idea for extending this in v2, but might as well do it in this PR if you think it's something we want.